File: apps/api/src/application/admin/admin-visit-type-service.ts
import type { AdminPayload } from '../support/auth-service.js';
import type { ActivityLogger } from '../support/activity-logger.js';
import type { VisitTypeRepository } from '../../infrastructure/repositories/visit-type-repository.js';
import { HttpError } from '../../rest/http-error.js';

export class AdminVisitTypeService {
	constructor(
		private readonly visitTypes: VisitTypeRepository,
		private readonly activity: ActivityLogger
	) {}

	list() {
		return this.visitTypes.listAdminWithMacroName();
	}

	async create(
		input: { name: string; description?: string; macroAreaId: number },
		admin: AdminPayload
	) {
		const id = await this.visitTypes.insert({
			name: input.name.trim(),
			description: input.description?.trim() || null,
			macroAreaId: input.macroAreaId
		});
		await this.activity.log({
			adminUserId: Number(admin.sub),
			action: 'visit_type.create',
			entityType: 'visit_type',
			entityId: String(id)
		});
		return { id };
	}

	async update(
		input: { id: number; name: string; description?: string; macroAreaId: number },
		admin: AdminPayload
	) {
		await this.visitTypes.update(input.id, {
			name: input.name.trim(),
			description: input.description?.trim() || null,
			macroAreaId: input.macroAreaId
		});
		await this.activity.log({
			adminUserId: Number(admin.sub),
			action: 'visit_type.update',
			entityType: 'visit_type',
			entityId: String(input.id)
		});
		return { ok: true as const };
	}

	async delete_(input: { id: number }, admin: AdminPayload) {
		const [bc, rc] = await Promise.all([
			this.visitTypes.countBookingsByVisitTypeId(input.id),
			this.visitTypes.countReportsByVisitTypeId(input.id)
		]);
		if (bc + rc > 0) {
			throw new HttpError(412, 'Tipologia collegata a prenotazioni o referti');
		}
		await this.visitTypes.deleteById(input.id);
		await this.activity.log({
			adminUserId: Number(admin.sub),
			action: 'visit_type.delete',
			entityType: 'visit_type',
			entityId: String(input.id)
		});
		return { ok: true as const };
	}
}

File: apps/python-api/src/yourmedportal_api/application/admin/admin_visit_type_service.py
from __future__ import annotations

from yourmedportal_api.application.support.activity_logger import ActivityLogger
from yourmedportal_api.application.support.auth_service import AdminPayload
from yourmedportal_api.infrastructure.repositories.visit_type_repository import VisitTypeRepository
from yourmedportal_api.rest.http_error import HttpError


class AdminVisitTypeService:
    def __init__(self, visit_types: VisitTypeRepository, activity: ActivityLogger) -> None:
        self._vt = visit_types
        self._activity = activity

    def list(self) -> list[dict]:
        return self._vt.list_admin_with_macro_name()

    def create(
        self, name: str, description: str | None, macro_area_id: int, admin: AdminPayload
    ) -> dict:
        new_id = self._vt.insert(
            name.strip(), description.strip() if description and description.strip() else None, macro_area_id
        )
        self._activity.log(
            admin_user_id=int(admin.sub),
            action="visit_type.create",
            entity_type="visit_type",
            entity_id=str(new_id),
        )
        return {"id": new_id}

    def update(
        self,
        id_: int,
        name: str,
        description: str | None,
        macro_area_id: int,
        admin: AdminPayload,
    ) -> dict:
        self._vt.update(
            id_,
            name.strip(),
            description.strip() if description and description.strip() else None,
            macro_area_id,
        )
        self._activity.log(
            admin_user_id=int(admin.sub),
            action="visit_type.update",
            entity_type="visit_type",
            entity_id=str(id_),
        )
        return {"ok": True}

    def delete(self, id_: int, admin: AdminPayload) -> dict:
        bc = self._vt.count_bookings_by_visit_type_id(id_)
        rc = self._vt.count_reports_by_visit_type_id(id_)
        if bc + rc > 0:
            raise HttpError(412, "Tipologia collegata a prenotazioni o referti")
        self._vt.delete_by_id(id_)
        self._activity.log(
            admin_user_id=int(admin.sub),
            action="visit_type.delete",
            entity_type="visit_type",
            entity_id=str(id_),
        )
        return {"ok": True}
